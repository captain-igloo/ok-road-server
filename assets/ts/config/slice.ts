import { createSlice } from '@reduxjs/toolkit';
import type { PayloadAction } from '@reduxjs/toolkit';

export interface ConfigState {
    demo: boolean;
    map: {
        center: {
            lat: number;
            lng: number;
        };
        zoom: number;
    };
    maxResults: number;
    speedLimitTilesUrl?: string;
}

const initialState: ConfigState = {
    demo: false,
    map: {
        center: {
            lat: -41,
            lng: 174,
        },
        zoom: 5,
    },
    maxResults: 1000,
    speedLimitTilesUrl: undefined,
};

export const configSlice = createSlice({
    name: 'config',
    initialState,
    reducers: {
        setConfig: (state, action: PayloadAction<ConfigState>) => {
            state.demo = action.payload.demo;
            state.map = action.payload.map;
            state.maxResults = action.payload.maxResults;
            state.speedLimitTilesUrl = action.payload.speedLimitTilesUrl;
        },
    },
});

export const { setConfig } = configSlice.actions;

export default configSlice.reducer;
