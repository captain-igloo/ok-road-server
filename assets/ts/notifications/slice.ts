import { createSlice } from '@reduxjs/toolkit';
import type { PayloadAction } from '@reduxjs/toolkit';

import type { AppDispatch } from '../store';

export interface NotificationsState {
    notificationCount: number;
    notifications: { [key: string]: string };
}

const initialState: NotificationsState = {
    notificationCount: 0,
    notifications: {},
};

export const notificationsSlice = createSlice({
    name: 'notifications',
    initialState,
    reducers: {
        addNotification: (state, action: PayloadAction<string>) => {
            state.notificationCount += 1;
            state.notifications[state.notificationCount] = action.payload;
        },
        removeNotification: (state, action: PayloadAction<number | undefined>) => {
            const notificationId = action.payload === undefined
                ? Object.keys(state.notifications)[0]
                : action.payload;
            delete state.notifications[notificationId];
        },
    },
});

export const addNotification = (notification: string) => (dispatch: AppDispatch) => {
    dispatch(notificationsSlice.actions.addNotification(notification));
    setTimeout(() => {
        dispatch(notificationsSlice.actions.removeNotification());
    }, 3000);
};

export const { removeNotification } = notificationsSlice.actions;

export default notificationsSlice.reducer;
