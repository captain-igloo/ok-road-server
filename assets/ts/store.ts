import { combineReducers, configureStore } from '@reduxjs/toolkit';

import configReducer from './config/slice';
import friendsReducer from './friends/slice';
import mapReducer from './map/slice';
import notificationsReducer from './notifications/slice';

const rootReducer = combineReducers({
    config: configReducer,
    friends: friendsReducer,
    notifications: notificationsReducer,
    map: mapReducer,
});

export const setupStore = (preloadedState?: Partial<RootState>) => configureStore({
    reducer: rootReducer,
    preloadedState,
});

export type AppStore = ReturnType<typeof setupStore>;

export type RootState = ReturnType<typeof rootReducer>;

export type AppDispatch = AppStore['dispatch'];
