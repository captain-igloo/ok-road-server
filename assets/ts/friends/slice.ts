import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import type { PayloadAction } from '@reduxjs/toolkit';

import { addNotification } from '../notifications/slice';
import type { AppDispatch, RootState } from '../store';

interface Friend {
    email: string;
    full_name: string;
    id: number;
    username: string;
}

export interface FriendsState {
    friends: Friend[];
    show: boolean;
}

const initialState: FriendsState = {
    friends: [],
    show: false,
};

export const friendsSlice = createSlice({
    name: 'friends',
    initialState,
    reducers: {
        showFriends: (state, action: PayloadAction<boolean>) => {
            state.show = action.payload;
        },
    },
    extraReducers: (builder) => {
        builder.addCase(fetchFriends.fulfilled, (state, action) => {
            state.friends = action.payload;
        });
    },
});

export const fetchFriends = createAsyncThunk<Friend[], void, { dispatch: AppDispatch; state: RootState; }>(
    'friends/fetchFriendsStatus',
    async (_, { dispatch }) => {
        const url = '/api/friends';
        const response = await fetch(url);
        if (!response.ok) {
            dispatch(addNotification('Failed to get friends'));
            throw new Error(`Failed to fetch: ${url}`);
        }
        return response.json();
    },
);

export const addFriend = createAsyncThunk<Friend, string, { dispatch: AppDispatch, state: RootState }>(
    'friends/addFriendStatus',
    async (username: string, { dispatch }) => {
        const url = '/api/friends';
        const response = await fetch(url, {
            body: JSON.stringify({ username }),
            headers: {
                accept: 'application/json',
                'content-type': 'application/json',
            },
            method: 'post',
        });
        if (!response.ok) {
            let message = 'Failed to add friend';
            if (response.status === 404) {
                message = 'User not found';
            } else if (response.status === 409) {
                message = `You are already friends with ${username}`;
            }
            dispatch(addNotification(message));
            throw new Error(`Failed to fetch: ${url}`);
        }
        dispatch(fetchFriends());
        return response.json();
    },
);

export const deleteFriend = createAsyncThunk<void, number, { dispatch: AppDispatch; state: RootState; }>(
    'friends/deleteFriendStatus',
    async (id: number, { dispatch }) => {
        const url = `/api/friends/${id}`;
        const response = await fetch(url, {
            method: 'delete',
        });
        if (!response.ok) {
            dispatch(addNotification('Failed to delete friend'));
            throw new Error(`Failed to fetch: ${url}`);
        }
        await dispatch(fetchFriends());
    },
);

export const showFriends = (show: boolean) => (dispatch: AppDispatch) => {
    dispatch(friendsSlice.actions.showFriends(show));
    if (show) {
        dispatch(fetchFriends());
    }
};

export default friendsSlice.reducer;
