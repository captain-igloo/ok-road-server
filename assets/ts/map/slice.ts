import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import type { PayloadAction } from '@reduxjs/toolkit';
import { LatLng, LatLngBounds } from 'leaflet';

import { addNotification } from '../notifications/slice';
import type { AppDispatch, RootState } from '../store';

export interface Feature {
    coordinates: number[];
    id: number;
    insertTimestamp: number;
    speedLimit?: {
        description: string;
        speedLimit: number;
    };
    timestamp: number;
    velocity: number;
}

export interface Device {
    description: string;
    id: number;
    name: string;
    username: string;
}

export interface User {
    fullName: string;
    username: string;
}

export interface Tooltip {
    position?: {
        lat: number;
        lng: number;
    };
    text: string[];
}

export interface MapState {
    bounds?: [[number, number], [number, number]];
    devices: Device[];
    features: Feature[];
    fromDate: number;
    highlightedLocation?: number;
    last24Hours: boolean;
    notificationCount: number;
    notifications: { [key: string]: string };
    refreshInProgress: boolean;
    selectedDevice?: number;
    showSpeedLimitAreas: boolean;
    toDate: number;
    tooltip: Tooltip;
    user: User;
}

const initialState: MapState = {
    bounds: undefined,
    devices: [],
    features: [],
    fromDate: Date.now() - (60 * 60 * 24 * 1000),
    highlightedLocation: undefined,
    last24Hours: true,
    notificationCount: 0,
    notifications: {},
    refreshInProgress: false,
    selectedDevice: undefined,
    showSpeedLimitAreas: false,
    toDate: Date.now(),
    tooltip: {
        position: undefined,
        text: [],
    },
    user: {
        fullName: '',
        username: '',
    },
};

export const fetchDevices = createAsyncThunk<
Device[],
void,
{
    dispatch: AppDispatch;
    state: RootState;
}>(
    'map/fetchDevicesStatus',
    async (_, { dispatch }) => {
        const url = '/api/devices';
        const response = await fetch(url);
        if (!response.ok) {
            dispatch(addNotification('Failed to fetch devices'));
            throw new Error(`Failed to fetch: ${url}`);
        }
        return response.json();
    },
);

export const fetchLocations = createAsyncThunk<{
    id: number;
    insert_timestamp: number;
    location: {
        x: number;
        y: number;
    };
    speed?: number;
    speed_limit?: {
        description: string;
        speed_limit: number;
    };
    timestamp: number;
}[], void, { dispatch: AppDispatch; state: RootState }>(
    'map/fetchLocationsStatus',
    async (_, { dispatch, getState }) => {
        const { fromDate } = getState().map;
        const { toDate } = getState().map;
        const { selectedDevice } = getState().map;
        if (selectedDevice) {
            dispatch(mapSlice.actions.setRefreshInProgress(true));
            const url = `/api/locations?device=${selectedDevice}`
                + `&from=${(new Date(fromDate)).toISOString()}&to=${(new Date(toDate)).toISOString()}`;
            const response = await fetch(url, {
                redirect: 'manual',
            });
            dispatch(mapSlice.actions.setRefreshInProgress(false));
            if (!response.ok) {
                dispatch(addNotification('Failed to fetch locations'));
                throw new Error(`Failed to fetch ${url}`);
            }
            const features = await response.json();
            if (features.length >= getState().config.maxResults) {
                dispatch(addNotification(`Results have been limited to ${getState().config.maxResults}`));
            }
            return features;
        }
        return [];
    },
);

export const mapSlice = createSlice({
    name: 'map',
    initialState,
    reducers: {
        highlightLocation: (state, action: PayloadAction<number | undefined>) => {
            state.highlightedLocation = action.payload;
        },
        setRefreshInProgress: (state, action: PayloadAction<boolean>) => {
            state.refreshInProgress = action.payload;
        },
        selectDevice: (state, action: PayloadAction<number>) => {
            state.selectedDevice = action.payload;
        },
        setFromDate: (state, action: PayloadAction<number>) => {
            state.fromDate = action.payload;
        },
        setLast24Hours: (state, action: PayloadAction<boolean>) => {
            state.last24Hours = action.payload;
        },
        setShowSpeedLimitAreas: (state, action: PayloadAction<boolean>) => {
            state.showSpeedLimitAreas = action.payload;
        },
        setToDate: (state, action: PayloadAction<number>) => {
            state.toDate = action.payload;
        },
        setTooltip: (state, action: PayloadAction<{ position?: { lat: number; lng: number }; text: string[] }>) => {
            state.tooltip = action.payload;
        },
        setUser: (state, action: PayloadAction<User>) => {
            state.user = action.payload;
        },
    },
    extraReducers: (builder) => {
        builder.addCase(fetchDevices.fulfilled, (state, action) => {
            state.devices = action.payload;
            if (action.payload.length > 0) {
                state.selectedDevice = action.payload[0].id;
            }
        });
        builder.addCase(fetchLocations.fulfilled, (state, action) => {
            let bounds: LatLngBounds | undefined;
            state.features = [];
            action.payload.forEach((feature) => {
                const point = new LatLng(feature.location.y, feature.location.x);
                if (!bounds) {
                    bounds = new LatLngBounds(point, point);
                } else {
                    bounds.extend(point);
                }
                let speedLimit;
                if (feature.speed_limit) {
                    speedLimit = {
                        description: feature.speed_limit.description,
                        speedLimit: feature.speed_limit.speed_limit,
                    };
                }
                state.features.push({
                    coordinates: [feature.location.x, feature.location.y],
                    id: feature.id,
                    insertTimestamp: feature.insert_timestamp,
                    speedLimit,
                    timestamp: feature.timestamp,
                    velocity: feature.speed || 0,
                });
            });
            if (bounds) {
                state.bounds = [
                    [bounds.getSouthWest().lat, bounds.getSouthWest().lng],
                    [bounds.getNorthEast().lat, bounds.getNorthEast().lng],
                ];
            }
        });
    },
});

let refreshTimeout: NodeJS.Timeout;

export const scheduleRefresh = () => (dispatch: AppDispatch) => {
    clearTimeout(refreshTimeout);
    refreshTimeout = setTimeout(() => {
        dispatch(mapSlice.actions.setFromDate(Date.now() - (60 * 60 * 24 * 1000)));
        dispatch(mapSlice.actions.setToDate(Date.now()));
        dispatch(fetchLocations());
        dispatch(scheduleRefresh());
    }, 300000);
};

export const setFromDate = (fromDate: number) => (dispatch: AppDispatch) => {
    dispatch(mapSlice.actions.setFromDate(fromDate));
    dispatch(fetchLocations());
};

export const setToDate = (toDate: number) => (dispatch: AppDispatch) => {
    dispatch(mapSlice.actions.setToDate(toDate));
    dispatch(fetchLocations());
};

export const selectDevice = (deviceId: number) => (dispatch: AppDispatch) => {
    dispatch(mapSlice.actions.selectDevice(deviceId));
    dispatch(fetchLocations());
};

export const setLast24Hours = (last24Hours: boolean) => (dispatch: AppDispatch) => {
    dispatch(mapSlice.actions.setLast24Hours(last24Hours));
    if (last24Hours) {
        dispatch(scheduleRefresh());
    } else {
        clearTimeout(refreshTimeout);
    }
};

export const setShowSpeedLimitAreas = (showSpeedLimitAreas: boolean) => (dispatch: AppDispatch) => {
    if (!showSpeedLimitAreas) {
        dispatch(mapSlice.actions.setTooltip({ position: undefined, text: [] }));
    }
    dispatch(mapSlice.actions.setShowSpeedLimitAreas(showSpeedLimitAreas));
};

export const {
    highlightLocation,
    setTooltip,
    setUser,
} = mapSlice.actions;

export default mapSlice.reducer;
