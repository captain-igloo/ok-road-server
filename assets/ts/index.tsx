import * as Sentry from '@sentry/browser';
import * as React from 'react';
import Container from 'react-bootstrap/Container';
import * as ReactDOM from 'react-dom/client';
import { Provider } from 'react-redux';

import Header from './Header';
import Important from './Important';
import { setupStore } from './store';

const element = document.getElementById('app');
if (element) {
    const dataParams = element.getAttribute('data-params');
    let configuration;
    if (dataParams) {
        configuration = JSON.parse(dataParams);
        if (configuration.sentryDsn) {
            Sentry.init({
                dsn: configuration.sentryDsn,
            });
        }
    }
    const root = ReactDOM.createRoot(element);
    root.render(
        <Provider store={setupStore()}>
            <Header user={configuration.user} />
            <Container style={{ paddingTop: '5px' }}>
                <h1>Monitor Young Drivers</h1>
                <p>
                    Install an app on their device and keep track of:
                </p>
                <ul>
                    <li>Where they are</li>
                    <li>How fast they are driving</li>
                    <li>Their speed in relation to the speed limit</li>
                </ul>
                <Important>
                    Speed limit data is currently only available in New Zealand.
                </Important>
                <h2>Getting Started</h2>
                <ol>
                    <li>
                        Click
                        {' '}
                        <a href="/register">here</a>
                        {' '}
                        to register with OK Road.  Remember your username and password because you
                        will need it in step 3.
                    </li>
                    <li>
                        Install
                        {' '}
                        <a href="https://owntracks.org">Owntracks</a>
                        {' '}
                        on the device you wish to monitor.
                    </li>
                    <li>
                        Configure Owntracks to publish to OK Road:
                        <ul>
                            <li>Set Preferences / Connection / Endpoint / Host to "192.9.176.238".</li>
                            <li>Set Preferences / Connection / Endpoint / Port to "1883".</li>
                            <li>Set Preferences / Connection / Credentials / Username to your username.</li>
                            <li>Set Preferences / Connection / Credentials / Password to your password.</li>
                        </ul>
                    </li>
                    <li>
                        Make sure that Owntracks has permission to use location services:
                        <ul>
                            <li>Set Settings / Location / Owntracks / Location access to "Allow all the time".</li>
                            <li>Set Settings / Location / Owntracks / Use precise location to true.</li>
                        </ul>
                    </li>
                </ol>
                <img alt="Screenshot" src="/img/okroad.png" />
            </Container>
        </Provider>,
    );
}
