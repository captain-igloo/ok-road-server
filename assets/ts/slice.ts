import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import type { PayloadAction } from '@reduxjs/toolkit';
import { LatLng, LatLngBounds } from 'leaflet';

import { AppDispatch, RootState } from './store';

export interface Feature {
    coordinates: number[];
    id: number;
    speedLimit?: {
        description: string;
        speedLimit: number;
    };
    timestamp: number;
    velocity: number;
};

export interface Device {
    description: string;
    id: number;
    name: string;
}

export interface OkRoadState {
    bounds?: [[number, number], [number, number]];
    devices: Device[];
    features: Feature[];
    fromDate: number;
    highlightedLocation?: number;
    selectedDevice?: number;
    showRecent: boolean;
    toDate: number;
    user: {
        fullName: string;
        username: string;
    }
    value: number;
}

const initialState: OkRoadState = {
    bounds: undefined,
    devices: [],
    features: [],
    fromDate: Date.now() - (60 * 60 * 24 * 1000),
    highlightedLocation: undefined,
    selectedDevice: undefined,
    showRecent: true,
    toDate: Date.now(),
    user: {
        fullName: '',
        username: '',
    },
    value: 0,
}

export const fetchDevices = createAsyncThunk<{
    description: string;
    id: number;
    name: string;
}[], void, {state: RootState}>(
    'okroad/fetchDevicesStatus',
    async () => {
        const url = '/api/devices';
        const response = await fetch(url);
        if (!response.ok) {
            throw new Error(`Failed to fetch: ${url}`);
        }
        return await response.json();
    },
);

export const fetchLocations = createAsyncThunk<{
    id: number;
    location: {
        x: number;
        y: number;
    };
    speed?: number;
    speed_limit?: {
        description: string;
        speed_limit: number;
    };
    timestamp: number;
}[], void, {state: RootState}>(
    'okroad/fetchLocationsStatus',
    async (_, { getState }) => {
        let fromDate: number;
        let toDate: number;
        if (getState().okRoad.showRecent) {
            fromDate = Date.now() - (60 * 60 * 24 * 1000);
            toDate = Date.now();
        } else {
            fromDate = getState().okRoad.fromDate;
            toDate = getState().okRoad.toDate;
        }

        const response = await fetch(`/api/locations?device=a14&from=${(new Date(fromDate)).toISOString()}&to=${(new Date(toDate)).toISOString()}`);
        if (!response.ok) {
            throw new Error('Failed to fetch');
        }
        return await response.json();
    },
);

export const okRoadSlice = createSlice({
    name: 'okRoad',
    initialState,
    reducers: {
        highlightLocation: (state, action: PayloadAction<number | undefined>) => {
            state.highlightedLocation = action.payload;
        },
        setFromDate: (state, action: PayloadAction<number>) => {
            state.fromDate = action.payload;  
        },
        setShowRecent: (state, action: PayloadAction<boolean>) => {
            state.showRecent = action.payload;
        },
        setToDate: (state, action: PayloadAction<number>) => {
            state.toDate = action.payload;
        },
        setUser: (state, action: PayloadAction<{fullName: string}>) => {
            state.user.fullName = action.payload.fullName;
        },
    },
    extraReducers: (builder) => {
        builder.addCase(fetchDevices.fulfilled, (state, action) => {
            state.devices = action.payload;
        });
        builder.addCase(fetchLocations.fulfilled, (state, action) => {
            let bounds: LatLngBounds | undefined;
            state.features = [];
            action.payload.forEach((feature) => {
                const point = new LatLng(feature.location.y, feature.location.x);
                if (!bounds) {
                    bounds = new LatLngBounds(point, point);
                } else {
                    bounds.extend(point);
                }
                let speedLimit;
                if (feature.speed_limit) {
                    speedLimit = {
                        description: feature.speed_limit.description,
                        speedLimit: feature.speed_limit.speed_limit,
                    };
                }
                state.features.push({
                    coordinates: [feature.location.x, feature.location.y],
                    id: feature.id,
                    speedLimit,
                    timestamp: feature.timestamp,
                    velocity: feature.speed || 0,
                });
            });
            if (bounds) {
                state.bounds = [
                    [bounds.getSouthWest().lat, bounds.getSouthWest().lng],
                    [bounds.getNorthEast().lat, bounds.getNorthEast().lng],
                ];
            }
        });
    },
});

export const setFromDate = (fromDate: number) => (dispatch: AppDispatch) => {
    dispatch(okRoadSlice.actions.setFromDate(fromDate));
    dispatch(fetchLocations());
};

export const setToDate = (toDate: number) => (dispatch: AppDispatch) => {
    dispatch(okRoadSlice.actions.setToDate(toDate));
    dispatch(fetchLocations());  
};

export const setShowRecent = (showRecent: boolean) => (dispatch: AppDispatch) => {
    dispatch(okRoadSlice.actions.setShowRecent(showRecent));
    dispatch(fetchLocations());  
};

export const { highlightLocation, setUser } = okRoadSlice.actions;

export default okRoadSlice.reducer;
