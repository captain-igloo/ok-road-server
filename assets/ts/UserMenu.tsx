import { faGlobe } from '@fortawesome/free-solid-svg-icons/faGlobe';
import { faRightFromBracket } from '@fortawesome/free-solid-svg-icons/faRightFromBracket';
import { faRightToBracket } from '@fortawesome/free-solid-svg-icons/faRightToBracket';
import { faUser } from '@fortawesome/free-solid-svg-icons/faUser';
import { faUserGroup } from '@fortawesome/free-solid-svg-icons/faUserGroup';
import { faUserPlus } from '@fortawesome/free-solid-svg-icons/faUserPlus';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import * as React from 'react';
import Button from 'react-bootstrap/Button';
import ButtonGroup from 'react-bootstrap/ButtonGroup';
import Dropdown from 'react-bootstrap/Dropdown';
import { useDispatch } from 'react-redux';

import Friends from './friends/Friends';
import { showFriends } from './friends/slice';
import { AppDispatch } from './store';

interface Props {
    showMap?: boolean;
    showRegister?: boolean;
    showSignIn?: boolean;
    user?: {
        fullName: string;
    };
}

const useAppDispatch: () => AppDispatch = useDispatch;

export default function UserMenu(props: Props) {
    const {
        showMap,
        showRegister,
        showSignIn,
        user,
    } = props;

    if (user === undefined) {
        let registerButton;
        if (showRegister !== false) {
            registerButton = (
                <Button href="/register" variant="outline">
                    <FontAwesomeIcon icon={faUserPlus} />
                    <span className="display-small">
                        &nbsp;
                        Register
                    </span>
                </Button>
            );
        }

        let signInButton;
        if (showSignIn !== false) {
            signInButton = (
                <Button href="/login" variant="outline">
                    <FontAwesomeIcon icon={faRightToBracket} />
                    <span className="display-small">
                        &nbsp;
                        Sign in
                    </span>
                </Button>
            );
        }

        return (
            <ButtonGroup>
                {registerButton}
                {signInButton}
            </ButtonGroup>
        );
    }

    let mapButton;
    if (showMap !== false) {
        mapButton = (
            <Button href="/map" variant="outline">
                <FontAwesomeIcon icon={faGlobe} />
                <span className="display-small">
                    &nbsp;
                    Map
                </span>
            </Button>
        );
    }

    const dispatch = useAppDispatch();

    return (
        <>
            {mapButton}
            <Dropdown>
                <Dropdown.Toggle variant="outline">
                    <FontAwesomeIcon icon={faUser} />
                    <span className="display-small">
                        &nbsp;
                        {user.fullName}
                    </span>
                </Dropdown.Toggle>
                <Dropdown.Menu>
                    <Dropdown.Item onClick={() => { dispatch(showFriends(true)); }}>
                        <FontAwesomeIcon icon={faUserGroup} />
                        &nbsp;
                        Friends
                    </Dropdown.Item>
                    <Dropdown.Item href="/logout">
                        <FontAwesomeIcon icon={faRightFromBracket} />
                        &nbsp;
                        Logout
                    </Dropdown.Item>
                </Dropdown.Menu>
            </Dropdown>
            <Friends />
        </>
    );
}
