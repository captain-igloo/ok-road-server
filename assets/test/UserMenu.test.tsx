import {
    describe,
    expect,
    jest,
    test,
} from '@jest/globals';
import { act, fireEvent, render } from '@testing-library/react';
import * as React from 'react';
import { Provider } from 'react-redux';

import * as friendsSlice from '../ts/friends/slice';
import { setupStore } from '../ts/store';
import UserMenu from '../ts/UserMenu';

describe('UserMenu component', () => {
    test('UserMenu renders properly (not logged in)', () => {
        const { container } = render(<UserMenu showDemo showMap showRegister showSignIn />);
        expect(container).toMatchSnapshot();
    });

    test('UserMenu renders properly (logged in)', () => {
        const { container } = render(
            <Provider store={setupStore()}>
                <UserMenu showDemo showMap showRegister showSignIn user={{fullName: 'Test User'}}/>
            </Provider>
        );
        expect(container).toMatchSnapshot();
    });

    test('Click "Friends" should show friends', async () => {
        const showFriends = jest.spyOn(friendsSlice, 'showFriends');
        const { getByText } = render(
            <Provider store={setupStore()}>
                <UserMenu showDemo showMap showRegister showSignIn user={{fullName: 'Test User'}}/>
            </Provider>
        );
        await act(async () => {
            await fireEvent.click(getByText('Test User'));
        });
        await act(async () => {
            await fireEvent.click(getByText('Friends'));
        });
        expect(showFriends).toHaveBeenCalledWith(true);
        showFriends.mockRestore();
    });
});
