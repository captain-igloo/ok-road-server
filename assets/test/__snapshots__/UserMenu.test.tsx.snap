// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserMenu component UserMenu renders properly (logged in) 1`] = `
<div>
  <a
    class="btn btn-outline"
    href="/map"
    role="button"
    tabindex="0"
  >
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-globe "
      data-icon="globe"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M352 256c0 22.2-1.2 43.6-3.3 64l-185.3 0c-2.2-20.4-3.3-41.8-3.3-64s1.2-43.6 3.3-64l185.3 0c2.2 20.4 3.3 41.8 3.3 64zm28.8-64l123.1 0c5.3 20.5 8.1 41.9 8.1 64s-2.8 43.5-8.1 64l-123.1 0c2.1-20.6 3.2-42 3.2-64s-1.1-43.4-3.2-64zm112.6-32l-116.7 0c-10-63.9-29.8-117.4-55.3-151.6c78.3 20.7 142 77.5 171.9 151.6zm-149.1 0l-176.6 0c6.1-36.4 15.5-68.6 27-94.7c10.5-23.6 22.2-40.7 33.5-51.5C239.4 3.2 248.7 0 256 0s16.6 3.2 27.8 13.8c11.3 10.8 23 27.9 33.5 51.5c11.6 26 20.9 58.2 27 94.7zm-209 0L18.6 160C48.6 85.9 112.2 29.1 190.6 8.4C165.1 42.6 145.3 96.1 135.3 160zM8.1 192l123.1 0c-2.1 20.6-3.2 42-3.2 64s1.1 43.4 3.2 64L8.1 320C2.8 299.5 0 278.1 0 256s2.8-43.5 8.1-64zM194.7 446.6c-11.6-26-20.9-58.2-27-94.6l176.6 0c-6.1 36.4-15.5 68.6-27 94.6c-10.5 23.6-22.2 40.7-33.5 51.5C272.6 508.8 263.3 512 256 512s-16.6-3.2-27.8-13.8c-11.3-10.8-23-27.9-33.5-51.5zM135.3 352c10 63.9 29.8 117.4 55.3 151.6C112.2 482.9 48.6 426.1 18.6 352l116.7 0zm358.1 0c-30 74.1-93.6 130.9-171.9 151.6c25.5-34.2 45.2-87.7 55.3-151.6l116.7 0z"
        fill="currentColor"
      />
    </svg>
    <span
      class="display-small"
    >
        Map
    </span>
  </a>
  <div
    class="dropdown"
  >
    <button
      aria-expanded="false"
      class="dropdown-toggle btn btn-outline"
      id="react-aria-«r0»"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-user "
        data-icon="user"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M224 256A128 128 0 1 0 224 0a128 128 0 1 0 0 256zm-45.7 48C79.8 304 0 383.8 0 482.3C0 498.7 13.3 512 29.7 512l388.6 0c16.4 0 29.7-13.3 29.7-29.7C448 383.8 368.2 304 269.7 304l-91.4 0z"
          fill="currentColor"
        />
      </svg>
      <span
        class="display-small"
      >
         
        Test User
      </span>
    </button>
  </div>
</div>
`;

exports[`UserMenu component UserMenu renders properly (not logged in) 1`] = `
<div>
  <div
    class="btn-group"
    role="group"
  >
    <a
      class="btn btn-outline"
      href="/register"
      role="button"
      tabindex="0"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-user-plus "
        data-icon="user-plus"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 640 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M96 128a128 128 0 1 1 256 0A128 128 0 1 1 96 128zM0 482.3C0 383.8 79.8 304 178.3 304l91.4 0C368.2 304 448 383.8 448 482.3c0 16.4-13.3 29.7-29.7 29.7L29.7 512C13.3 512 0 498.7 0 482.3zM504 312l0-64-64 0c-13.3 0-24-10.7-24-24s10.7-24 24-24l64 0 0-64c0-13.3 10.7-24 24-24s24 10.7 24 24l0 64 64 0c13.3 0 24 10.7 24 24s-10.7 24-24 24l-64 0 0 64c0 13.3-10.7 24-24 24s-24-10.7-24-24z"
          fill="currentColor"
        />
      </svg>
      <span
        class="display-small"
      >
          Register
      </span>
    </a>
    <a
      class="btn btn-outline"
      href="/login"
      role="button"
      tabindex="0"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-right-to-bracket "
        data-icon="right-to-bracket"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M217.9 105.9L340.7 228.7c7.2 7.2 11.3 17.1 11.3 27.3s-4.1 20.1-11.3 27.3L217.9 406.1c-6.4 6.4-15 9.9-24 9.9c-18.7 0-33.9-15.2-33.9-33.9l0-62.1L32 320c-17.7 0-32-14.3-32-32l0-64c0-17.7 14.3-32 32-32l128 0 0-62.1c0-18.7 15.2-33.9 33.9-33.9c9 0 17.6 3.6 24 9.9zM352 416l64 0c17.7 0 32-14.3 32-32l0-256c0-17.7-14.3-32-32-32l-64 0c-17.7 0-32-14.3-32-32s14.3-32 32-32l64 0c53 0 96 43 96 96l0 256c0 53-43 96-96 96l-64 0c-17.7 0-32-14.3-32-32s14.3-32 32-32z"
          fill="currentColor"
        />
      </svg>
      <span
        class="display-small"
      >
          Sign in
      </span>
    </a>
  </div>
</div>
`;
